File: resources/demo/sample-solutions/valid-parentheses.py
"""
Valid Parentheses - Python Solutions
======================================
Problem: Determine if a string of brackets is valid.
"""

# ============================================================
# Solution 1: Stack (Standard)
# Time: O(n), Space: O(n)
# ============================================================
def is_valid(s):
    """
    Use a stack to track opening brackets.
    When we see a closing bracket, check if it matches the top of the stack.
    """
    stack = []
    mapping = {')': '(', '}': '{', ']': '['}

    for char in s:
        if char in mapping:
            # Closing bracket: check if it matches the top
            if not stack or stack[-1] != mapping[char]:
                return False
            stack.pop()
        else:
            # Opening bracket: push onto stack
            stack.append(char)

    # Valid only if all brackets were matched
    return len(stack) == 0


# ============================================================
# Solution 2: Replacement-based (clever but less efficient)
# Time: O(n^2) in worst case, Space: O(n)
# ============================================================
def is_valid_replace(s):
    """
    Repeatedly remove valid pairs until the string is empty or
    no more pairs can be removed.
    """
    while '()' in s or '{}' in s or '[]' in s:
        s = s.replace('()', '').replace('{}', '').replace('[]', '')
    return s == ''


# ============================================================
# Solution 3: Stack with early termination
# Time: O(n), Space: O(n)
# ============================================================
def is_valid_optimized(s):
    """Optimized with early termination: if remaining chars can't form pairs."""
    if len(s) % 2 != 0:
        return False  # Odd length can't be valid

    stack = []
    mapping = {')': '(', '}': '{', ']': '['}

    for char in s:
        if char in mapping:
            if not stack:
                return False  # No matching opening bracket
            if stack[-1] != mapping[char]:
                return False
            stack.pop()
        else:
            stack.append(char)

    return not stack


# ============================================================
# Test
# ============================================================
if __name__ == "__main__":
    assert is_valid("()") == True
    assert is_valid("()[]{}") == True
    assert is_valid("(]") == False
    assert is_valid("([)]") == False
    assert is_valid("{[]}") == True
    assert is_valid("") == True
    assert is_valid("(") == False
    assert is_valid(")") == False
    print("All tests passed!")

File: resources/demo/sample-solutions/valid-parentheses.js
/**
 * Valid Parentheses - JavaScript Solutions
 * =========================================
 * Problem: Determine if a string of brackets is valid.
 */

// ============================================================
// Solution 1: Stack (Standard)
// Time: O(n), Space: O(n)
// ============================================================
function isValid(s) {
  const stack = [];
  const mapping = { ')': '(', '}': '{', ']': '[' };

  for (const char of s) {
    if (char in mapping) {
      // Closing bracket: check if it matches the top
      if (stack.length === 0 || stack[stack.length - 1] !== mapping[char]) {
        return false;
      }
      stack.pop();
    } else {
      // Opening bracket: push onto stack
      stack.push(char);
    }
  }

  return stack.length === 0;
}

// ============================================================
// Solution 2: Stack with early termination
// Time: O(n), Space: O(n)
// ============================================================
function isValidOptimized(s) {
  if (s.length % 2 !== 0) return false;

  const stack = [];

  for (const char of s) {
    switch (char) {
      case '(': stack.push(')'); break;
      case '{': stack.push('}'); break;
      case '[': stack.push(']'); break;
      default:
        if (stack.length === 0 || stack.pop() !== char) {
          return false;
        }
    }
  }

  return stack.length === 0;
}

// ============================================================
// Test
// ============================================================
console.assert(isValid("()") === true);
console.assert(isValid("()[]{}") === true);
console.assert(isValid("(]") === false);
console.assert(isValid("([)]") === false);
console.assert(isValid("{[]}") === true);
console.assert(isValid("") === true);
console.log("All tests passed!");
